File: scripts/visual-qa.mjs
/**
 * LAYER 2 RUNNER — measures a real rendered page in a real browser.
 *
 * Usage: node scripts/visual-qa.mjs http://localhost:8080/s/<slug> [more urls]
 *
 * It loads each URL at every width Revora judges, runs the shared measurement
 * snippet from src/lib/builder/visual.ts, grades the numbers with the same code
 * the app uses, and prints the verdict. Nothing here invents a score: with no
 * measurements, the report says the site has not been checked.
 */
import { chromium } from "playwright";
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

const urls = process.argv.slice(2);
if (!urls.length) {
  console.error("Give at least one URL to check.");
  process.exit(1);
}

// The grader is TypeScript, so it is imported through the app's own tooling when
// available and otherwise evaluated from source via bun.
const { VIEWPORTS, MEASURE_SCRIPT, gradeVisual } = await import(
  pathToFileURL(new URL("../src/lib/builder/visual.ts", import.meta.url).pathname).href
).catch(() => {
  throw new Error(
    "Run this with bun so TypeScript sources can be imported: bun scripts/visual-qa.mjs <url>",
  );
});

void readFileSync;

const browser = await chromium.launch();
let failures = 0;

for (const url of urls) {
  const measurements = [];
  for (const width of VIEWPORTS) {
    const context = await browser.newContext({ viewport: { width, height: 1800 } });
    const page = await context.newPage();
    try {
      await page.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
      await page.waitForTimeout(300);
      measurements.push(await page.evaluate(MEASURE_SCRIPT));
    } catch (error) {
      console.error(`could not load ${url} at ${width}px:`, error.message);
    } finally {
      await context.close();
    }
  }

  const report = gradeVisual(measurements);
  console.log(`\n${url} — ${report.score}/100 ${report.passed ? "PASS" : "FAIL"}`);
  for (const finding of report.findings)
    console.log(`  [${finding.severity}] ${finding.width}px ${finding.key}: ${finding.detail}`);
  if (!report.passed) failures += 1;
}

await browser.close();
process.exit(failures ? 1 : 0);

File: scripts/visual-qa.py
"""LAYER 2 MEASURING — loads real pages in a real browser and records what it sees.

Usage:  python3 scripts/visual-qa.py <url> [more urls]

For every width Revora judges, it runs the shared measurement snippet from
src/lib/builder/visual.ts and writes the raw numbers to /tmp/visual-qa.json.
Grading is NOT done here: `bun scripts/visual-grade.ts` scores the same numbers
with the exact code the app uses, so a claim about how a page looks always comes
from one grader.
"""

import asyncio
import json
import re
import sys
from pathlib import Path

from playwright.async_api import async_playwright

ROOT = Path(__file__).resolve().parent.parent
SOURCE = (ROOT / "src/lib/builder/visual.ts").read_text()

WIDTHS = [int(n) for n in re.search(r"VIEWPORTS = \[([^\]]+)\]", SOURCE).group(1).split(",") if n.strip().isdigit()]
SNIPPET = SOURCE.split("export const MEASURE_SCRIPT = `", 1)[1].rsplit("`;", 1)[0]


async def main(urls):
    out = []
    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        for url in urls:
            page_results = []
            for width in WIDTHS:
                context = await browser.new_context(viewport={"width": width, "height": 1800})
                page = await context.new_page()
                try:
                    await page.goto(url, wait_until="domcontentloaded", timeout=30000)
                    await page.wait_for_timeout(400)
                    page_results.append(await page.evaluate(SNIPPET))
                except Exception as error:  # noqa: BLE001 - reported, never hidden
                    print(f"could not load {url} at {width}px: {error}", file=sys.stderr)
                finally:
                    await context.close()
            out.append({"url": url, "measurements": page_results})
        await browser.close()
    Path("/tmp/visual-qa.json").write_text(json.dumps(out))
    print(f"measured {sum(len(item['measurements']) for item in out)} viewport(s) -> /tmp/visual-qa.json")


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Give at least one URL to check.", file=sys.stderr)
        raise SystemExit(1)
    asyncio.run(main(sys.argv[1:]))
